src/pratt2.rs

```diff
@@ -0,0 +1,296 @@
+#![allow(missing_docs)]
+
+use super::*;
+
+trait Operator<'a, I, O, E>
+where
+    I: Input<'a>,
+    E: ParserExtra<'a, I>,
+{
+    type Op;
+    type OpParser: Parser<'a, I, Self::Op, E>;
+    const INFIX: bool = false;
+    const PREFIX: bool = false;
+
+    fn op_parser(&self) -> &Self::OpParser;
+    fn associativity(&self) -> Associativity;
+    fn fold_infix(&self, lhs: O, op: Self::Op, rhs: O) -> O {
+        unreachable!()
+    }
+    fn fold_prefix(&self, op: Self::Op, rhs: O) -> O {
+        unreachable!()
+    }
+}
+
+#[derive(Copy, Clone, PartialEq)]
+pub enum Associativity {
+    Left(u16),
+    Right(u16),
+}
+
+impl Associativity {
+    fn left_power(&self) -> u16 {
+        match self {
+            Self::Left(x) => x * 2,
+            Self::Right(x) => x * 2 + 1,
+        }
+    }
+
+    fn right_power(&self) -> u16 {
+        match self {
+            Self::Left(x) => x * 2 + 1,
+            Self::Right(x) => x * 2,
+        }
+    }
+}
```

Comment:
This is probably the bit I'm less confident on, I'm still a little bit iffy about how `Associativity` applies for prefix/postfix operators (my guess would be that it simply doesn't, and this can probably be boiled away into nothing). Would like to hear your thoughts here :)